### PR TITLE
Resolved Windows related issues

### PR DESCRIPTION
**Pull request by jgebal**
The changes resolve 2 issues:
1. oci8 library not incldued by Bundler on Windows
2. Timezone settings not working on windows

### CODE REVIEW

**Comment by jgebal** [comment]
This pull request addresses issue #82.
No new tests were added, as the existing tests were failing on Windows prior to the changes, so they document the fix

File: spec/plsql/schema_spec.rb [hunk 1]
```diff
@@ -86,7 +86,7 @@
   it "should not connect with wrong port number" do
     expect {
       plsql.connect! @username, @password, :host => @host, :port => 9999, :database => @database
-    }.to raise_error(/no listener|could not establish the connection/)
+    }.to raise_error(/ORA-12541/)
```
Comment (jgebal) [inline spec/plsql/schema_spec.rb]: The test was failing if Oracle locale set to different than English.
I assume it's better to use Oracle error numbers than strings, as the strings are language-specific

Comment (javornikolov) [inline spec/plsql/schema_spec.rb]: What about setting `NLS_LANG=AMERICAN_AMERICA.AL32UTF8` in spec_helper?

File: spec/spec_helper.rb [hunk 2]
```diff
@@ -112,7 +112,10 @@ def only(*whitelist)
...
-# which will be used to set session time zone
-ENV['TZ'] ||= 'Europe/Riga'
-# ENV['TZ'] ||= 'UTC'
+# Set Oracle Database session timezone
+# this setting needs to be in place to make sure that Oracle properly represents
+# the timestamp format on your local machine
+# Do not use 'OS_TZ' value or fixed offset values like '05:00', for daylight saving-enabled timezones
+# The setting cannot be derived directly from Windows operating system or ruby, as on the Windows machines,
+# timezone names are non-standard ones and are subject to translations to different languages.
+ENV['ORA_SDTZ'] ||= 'Europe/London'
```
Comment (javornikolov) [inline spec/spec_helper.rb]: Can we clean that `No new line at end of file` warning?

Comment (javornikolov) [inline spec/spec_helper.rb]: @jgebal - do you mean on Windows if you don't set ORA_SDTZ (or set it to OS_TZ), then `select sessiontimezone from dual` is not correct representation of your local time zone?

Right now on my environment some tests started failing... Maybe we should leave ORA_SDTZ unset by default (or set to OS_TZ) and override it just in case of need (broken OS or something). Could be via pulling some optional .gitignore-d file.

File: lib/plsql/connection.rb [hunk 3]
```diff
@@ -197,9 +197,8 @@ def session_id
       @session_id ||= select_first("SELECT TO_NUMBER(USERENV('SESSIONID')) FROM dual")[0]
     end
 
-    # Set time zone (default taken from TZ environment variable)
+    # Set time zone
     def set_time_zone(time_zone=nil)
-      time_zone ||= ENV['TZ']
```
Comment (javornikolov) [inline lib/plsql/connection.rb]: With such a change a call on [spec/plsql/connection_spec.rb#L10](https://github.com/rsim/ruby-plsql/blob/master/spec/plsql/connection_spec.rb#L10) becomes redundant. Maybe we can get rid of it?

I wonder about the default value: should we keep it `nil` or make it `local` instead?

File: spec/spec_helper.rb (hunk 2 above)
Comment (javornikolov) [inline spec/spec_helper.rb]: OK... Oracle by default (or with OS_TZ) is using non-DST enabled offset (works that way on Linux too). So seems we really need to override it if our OS timezone is DST-enabled.

Something else is flawed (maybe in ruby-oci8) that setting/getting timestamp is not symmetric but that's a subject of separate investigation (out of scope for this PR).

File: spec/plsql/schema_spec.rb (hunk 1 above)
Comment (jgebal) [inline spec/plsql/schema_spec.rb]: In fact, now that you mention it, it think it would be better not to set the ENV parameters at  all, or set them to some defaults, if they are not set.
That should go for:
NLS_LANG Oracle session language
ORA_SDTZ Oracle session timezone 
TZ ruby timezone parameter
So by default you would get:
```
TZ=UTC
ORA_SDTZ=UTC
NLS_LANG=AMERICAN_AMERICA.AL32UTF8
```
Anyway, users should be able to change those settings as they need, to setup any valid language and timezone on their environment.
I can update the `spec_helper.rb` to reflect that.
Would that be OK?

File: spec/spec_helper.rb (hunk 2 above)
Comment (jgebal) [inline spec/spec_helper.rb]: On my machine test were passing, when setting the `ORA_SDTZ='Europe/London'`
as my local timezone is actually "Europe/London".
Some test might have failed for you if you local timezone is different.
What I have noticed it that Ruby handles local timezone with DST more-less correct while Oracle uses fixed offset :(
That's why I've decided to fix only one side (Oracle), as setting timezone from Ruby will not do on Windows.
On windows tzutil /l gives following timezone names:
```
...
(UTC+10:00) Guam, Port Moresby
West Pacific Standard Time

(UTC+10:00) Hobart
Tasmania Standard Time

(UTC+10:00) Magadan
Magadan Standard Time

(UTC+10:00) Władywostok, Magadan (RTZ 9)
Vladivostok Standard Time
...
```
While Oracle valid timezone names `SELECT * FROM v$timezone_names` are like:
```
Africa/Algiers                                                   LMT
Africa/Algiers                                                   PMT
Africa/Algiers                                                   WET
Africa/Algiers                                                   WEST
Africa/Algiers                                                   CET
Africa/Algiers                                                   CEST
```

File: lib/plsql/connection.rb (hunk 3 above)
Comment (jgebal) [inline lib/plsql/connection.rb]: True, will do cleanup of those calls.
The nil as default doesn't seem to make sense anymore.
however LOCAL is a keyword not a TZ name, and again setting to LOCAL makes the dumb thing, setting timezone as a fixed offset.

File: spec/spec_helper.rb (hunk 2 above)
Comment (javornikolov) [inline spec/spec_helper.rb]: Yes, Oracle time zones names are Oracle-specific - that's not a Windows-specific issue I think. I agree about the approach with setting ORA_SDTZ.

Just I propose something like following in order to avoid the hassle of having local custom settings in version-controlled files:
* Add `spec/support/custom_config.rb` to `.gitignore`
* If anyone wants to set specific ORA_SDTZ: this should be done in spec/support/custom_config.rb
* We may provide commented sample `spec/support/custom_config.rb.sample` or/and maybe some defaults e.g. in `spec_helper`.

File: lib/plsql/connection.rb (hunk 3 above)
Comment (javornikolov) [inline lib/plsql/connection.rb]: `local` will reset to the initial value which has been at session start. So if we have `ORA_SDTZ=Europe/London` - then set to `local` will keep DST-awareness.

The question is not for plsql-spec own tests but for ruby-plsql API in general:
 - what behavior do we expect from `connection.set_time_zone`? Do we have default, do we allow passing nil value, what is it supposed to do?
 - There might be existing code relying on the old version (i.e. the TZ environment variable). We need to document our change somewhere and perhaps plan bumping up the gem version number accordingly.

Comment (jgebal) [inline lib/plsql/connection.rb]: So how about this one?
```ruby
    def set_time_zone(time_zone=ENV['ORA_SDTZ'])
      time_zone = time_zone ? "'#{time_zone}'" : 'LOCAL'
      exec('alter session set time_zone = ' + time_zone)
    end
```

Comment (javornikolov) [inline lib/plsql/connection.rb]: ORA_SDTZ may hold some values (OS_TZ, DB_TZ) which are not eligible for time_zone db parameter.

Let's think about possible options:
* Allow `nil` but do nothing (like your initial version)
* Or raise an exception (e.g. by passing it through and let Oracle raise error)
* Overwrite to mean 'local'
* Overwrite with ENV['TZ'] environment var - as it used to be. (Put a remark that this is not encouraged since tz names may be different in Oracle).
* Overwrite with ENV['RUBY_PLSQL_TZ']

File: spec/plsql/schema_spec.rb (hunk 1 above)
Comment (javornikolov) [inline spec/plsql/schema_spec.rb]: `+    }.to raise_error(/ORA-12541/)`
This breaks the test when using Jruby.

Comment (javornikolov) [inline spec/plsql/schema_spec.rb]: > In fact, now that you mention it, it think it would be better not to set the ENV parameters at all, or set them to some defaults, if they are not set.
> That should go for:
> NLS_LANG Oracle session language
> ORA_SDTZ Oracle session timezone 
> TZ ruby timezone parameter
> So by default you would get:
> 
>> TZ=UTC
>> ORA_SDTZ=UTC
>> NLS_LANG=AMERICAN_AMERICA.AL32UTF8
> Anyway, users should be able to change those settings as they need, to setup any valid language and timezone on their environment.
> I can update the spec_helper.rb to reflect that.
> Would that be OK?

* Sounds OK at first glimpse. I'm not sure about the TZ=UTC, maybe it would be better to use e.g. `Europe/Riga` as default - otherwise tests won't be able to capture a buggy DST-sensitive behavior.
* Let's say a user wants to save the local custom settings (to avoid typing them each time): what would be the best: a .gitignored file (like the idea with `custom_settings.rb`); or some external script which to remember to invoke (we anyway may need that for ORACLE_HOME, PATH, etc.)

Maybe it's best to keep customization (of TZ, etc) scope of impact as small as possible - e.g. for ruby-plsql rspec tests only, but not for my shell. For my shell - I may have my own preferences for local time, language, and so on.

**Comment by jgebal** [comment]
About the way to set timezone in Oracle.
I'd like to keep it as simple as possible.
Looking at the possible usage scenarios I see the following cases:

* ruby-plsql used as a server-side with connection pool for Oracle. We should provide an easy way to let Oracle know what is the client's (like browser) time zone.
This can be achieved now by calling .set_time_zone.
In this scenario there is no valuable case of calling the set_time_zone without parameters.
In this scenario both ruby and oracle timezone should be adjusted for the duration of the call (probably).

* ruby-plsql used as a library on under ruby-plsql-spec for Unit Testing Oracle code (my use case scenario). When used, I'd expect it to always set the session time zone to the callers (ruby) side.
That is almost good without any settings, except that daylight saving settings on caller are not preserved.
In this case, we only need to make sure that we fix the bad default settings  Oracle side.
This can be done by documenting (you need to setup your env properly) or by implementing a feature that allows to do that.

I assume there is no real case where we deliberately want to have different time zone set on ruby and on oracle session.

Having that said I wonder if the change I've introduced is actually needed.
Maybe an update to readme.md would be sufficient.

**Comment by jgebal** [comment]
Hmmm
Seems like it is not about the TZ names after all in Windows.
It is about the fact that assignments to ENV['TZ'] are simply ignored on Windows:
https://www.ruby-forum.com/topic/68570
So the only way to make sure that TZ is OK is to set it before you run your code on Windows not later.

Seems like I've touched something that's far more complex (buggy and messy) than it seems.

**Comment by javornikolov** [comment]
I think it was good to remove the ENV['TZ'] dependency from the code. We can write instructions that ORA_SDTZ needs to be set to the relevant local time.

In Oracle `TIMESTAMP` does not have time zone information. Maybe the proper way on Ruby side would be to discard timezone part and compare the rest.
In Java there is a zone-less type [LocalTime](https://docs.oracle.com/javase/8/docs/api/java/time/LocalTime.html), but I'm not sure if there is equivalent in Ruby.

**Comment by jgebal** [comment]
The updated README.md already documents this.
Can you summarize the things you'd like me to change?
Its hard to conclude from the discussions that were ongoing over the lines of code.
I'll be happy to have that pull request merged or rejected, whichever is more appropriate.
In the end, it seems we could have lived with the existing solution, just that the ORA_SDTZ needs to be set properly as an env variable and the Gemfile inclusion of oci8 had to be fixed for Windows.

Some general thought though.
_The spec_helper.rb became really messy and unreadable. Would be good to clean it up._

**Comment by javornikolov** [comment]
Yes, sure - let's sum up:
 - [ ] Revert the change with ` raise_error(/ORA-12541/)`. Set NLS_LANG instead.
 - [ ] The README mentions about `NLS_LANG`, but it doesn't give enough guidelines (it should be AMERICAN_AMERICA.AL32UTF8).
 - [ ] Similar for ORA_SDTZ - it's good to have a guideline that it's supposed to be a DST-aware time zone name.
 - [ ] Either live without a default or change `ENV['ORA_SDTZ'] ||= 'Europe/London'` to `Europe/Riga` (no other particular reasons than this is how it used to be before).
 - [ ] Consider adding something like `spec/support/custom_config.rb` to `.gitignore` to ease preserving and reusing some of the local custom settings (e.g. ORA_SDTZ, NLS_LANG). But maybe it's enough to without that and rely on environment variables.

**Comment by jgebal** [comment]
Thanks. Will do that and come back with a newer version.

**Comment by jgebal** [comment]
I will need to add one more thing.
Just checked on JRuby
and the session timezone on JDBC is not derived from ORA_SDTZ, so I will put setting the timezone on JDBC connection explicitly.

File: README.md [hunk 4]
```diff
@@ -138,6 +138,18 @@ If you are using JRuby then you need to download latest [Oracle JDBC driver](htt
 * in `JRUBY_HOME/lib` directory
 * or include path to JDBC driver jar file in Java `CLASSPATH`
 
+Make sure to setup the following Oracle-specific environment variables properly
+
+* [NLS_LANG](http://www.orafaq.com/wiki/NLS_LANG) - preferred value `NLS_LANG=AMERICAN_AMERICA.AL32UTF8`
+* [ORA_SDTZ](http://docs.oracle.com/cd/E18283_01/server.112/e10729/ch4datetime.htm#CBBEEAFB) The setting should point a machine timezone like: `ORA_SDTZ=Europe/Riga`, otherwise Oracle by default uses a Fixed-offset timezone (like `03:00`) that is not daylight saving (DST) aware, which will lead to wrong translations of the timestamp values between Ruby code (DTS-aware) and Oracle session (non-DST-aware).  
+* [ORACLE_HOME](http://www.orafaq.com/wiki/ORACLE_HOME)
+
+You may either alter your environment settings or set the values in file `spec\support\custom_config.rb`. Sample file `custom_config.rb.sample` shows how to do that.
```
Comment (javornikolov) [inline README.md]: Path separator should be `/` instead of `\` - in order to be consistent with other descriptions.

File: .gitignore [hunk 5]
```diff
@@ -11,3 +11,5 @@ sqlnet.log
 Gemfile.lock
 *.zip
 .vagrant
+.idea
+spec/support/custom_config.rb
```
Comment (javornikolov) [inline .gitignore]: Is it possible to fix that "No new line at end of file" warning? (If it's hard to do on Windows - let me know, I can push a fix for that).

Comment (jgebal) [inline .gitignore]: Not hard, just need to be more careful and change settings in my IDE.

*sent from mobile
2 cze 2015 18:57 "Yavor Nikolov" <notifications@github.com> napisał(a):

Comment (jgebal) [inline .gitignore]: fixed

File: README.md (hunk 4 above)
Comment (jgebal) [inline README.md]: fixed

File: README.md [hunk 6]
```diff
@@ -138,6 +138,18 @@ If you are using JRuby then you need to download latest [Oracle JDBC driver](htt
 * in `JRUBY_HOME/lib` directory
 * or include path to JDBC driver jar file in Java `CLASSPATH`
 
+Make sure to setup the following Oracle-specific environment variables properly
+
+* [NLS_LANG](http://www.orafaq.com/wiki/NLS_LANG) - preferred value `NLS_LANG=AMERICAN_AMERICA.AL32UTF8`
+* [ORA_SDTZ](http://docs.oracle.com/cd/E18283_01/server.112/e10729/ch4datetime.htm#CBBEEAFB) The setting should point a machine timezone like: `ORA_SDTZ=Europe/Riga`, otherwise Oracle by default uses a Fixed-offset timezone (like `03:00`) that is not daylight saving (DST) aware, which will lead to wrong translations of the timestamp values between Ruby code (DTS-aware) and Oracle session (non-DST-aware).
```
Comment (javornikolov) [inline README.md]: There is trailing white space at the end of this line: after `(non-DST-aware).`

File: README.md [hunk 7]
```diff
@@ -138,6 +138,18 @@ If you are using JRuby then you need to download latest [Oracle JDBC driver](htt
...
+Make sure to setup the following Oracle-specific environment variables properly
+
+* [NLS_LANG](http://www.orafaq.com/wiki/NLS_LANG) - preferred value `NLS_LANG=AMERICAN_AMERICA.AL32UTF8`
+* [ORA_SDTZ](http://docs.oracle.com/cd/E18283_01/server.112/e10729/ch4datetime.htm#CBBEEAFB) The setting should point a machine timezone like: `ORA_SDTZ=Europe/Riga`, otherwise Oracle by default uses a Fixed-offset timezone (like `03:00`) that is not daylight saving (DST) aware, which will lead to wrong translations of the timestamp values between Ruby code (DTS-aware) and Oracle session (non-DST-aware).  
+* [ORACLE_HOME](http://www.orafaq.com/wiki/ORACLE_HOME)
+
+You may either alter your environment settings or set the values in file `spec/support/custom_config.rb`. Sample file `custom_config.rb.sample` shows how to do that.
+
+
+Make sure you use correct version of Oracle client for database you're connecting to. Otherwise you may encounter TimeZone errors like [this](http://stackoverflow.com/questions/7678485/oracle-ora-01805-on-oracle-11g-database)
```
Comment (javornikolov) [inline README.md]: There is a trailing whitespace here too.

**Comment by javornikolov** [comment]
Hi @jgebal, thanks for your fixes! I just spotted a few trailing whitespaces in the README - it's good to remove them. Also, you may consider squashing some of the commits to get a cleaner commit history for this PR.

No other remarks. After sorting out the above points I think we can merge.

**Comment by jgebal** [comment]
@javornikolov 
The squash was a tricky part once changes pushed to remote.
That is done now.

**Comment by javornikolov** [comment]
Looks good now, I'm merging it... Thank you @jgebal for your contribution!
